executor/executor.go

```diff
@@ -0,0 +1,606 @@
+func (e *TableScanExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Execution Next interface.
+func (e *TableScanExec) Next() (*Row, error) {
+	if e.iter == nil {
+		txn, err := e.ctx.GetTxn(false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.iter, err = txn.Seek([]byte(e.t.FirstKey()))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	if !e.iter.Valid() || !strings.HasPrefix(e.iter.Key(), e.t.KeyPrefix()) {
+		return nil, nil
+	}
+	// TODO: check if lock valid
+	// the record layout in storage (key -> value):
+	// r1 -> lock-version
+	// r1_col1 -> r1 col1 value
+	// r1_col2 -> r1 col2 value
+	// r2 -> lock-version
+	// r2_col1 -> r2 col1 value
+	// r2_col2 -> r2 col2 value
+	// ...
+	rowKey := e.iter.Key()
+	handle, err := util.DecodeHandleFromRowKey(rowKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// TODO: we could just fetch mentioned columns' values
+	row := &Row{}
+	row.Data, err = e.t.Row(e.ctx, handle)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Set result fields value.
+	for i, v := range e.fields {
+		v.Expr.SetValue(row.Data[i])
+	}
+
+	// Put rowKey to the tail of record row
+	rke := &RowKeyEntry{
+		Tbl: e.t,
+		Key: rowKey,
+	}
+	row.RowKeys = append(row.RowKeys, rke)
+
+	rk := e.t.RecordKey(handle, nil)
+	err = kv.NextUntil(e.iter, util.RowKeyPrefixFilter(rk))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return row, nil
+}
+
+// Close implements plan.Plan Close interface.
+func (e *TableScanExec) Close() error {
+	if e.iter != nil {
+		e.iter.Close()
+		e.iter = nil
+	}
+	return nil
+}
+
+// IndexRangeExec represents an index range scan executor.
+type IndexRangeExec struct {
+	scan *IndexScanExec
+
+	// seekVal is different from lowVal, it is casted from lowVal and
+	// must be less than or equal to lowVal, used to seek the index.
+	lowVal      interface{}
+	lowExclude  bool
+	highVal     interface{}
+	highExclude bool
+
+	iter       kv.IndexIterator
+	skipLowCmp bool
+	finished   bool
+}
+
+// Fields implements Executor Fields interface.
+func (e *IndexRangeExec) Fields() []*ast.ResultField {
+	return e.scan.fields
+}
+
+// Next implements Executor Next interface.
+func (e *IndexRangeExec) Next() (*Row, error) {
+	if e.iter == nil {
+		var seekVal interface{}
+		var err error
+		if e.lowVal == plan.MinNotNullVal {
+			seekVal = []byte{}
+		} else {
+			seekVal, err = types.Convert(e.lowVal, &e.scan.valueType)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		txn, err := e.scan.ctx.GetTxn(false)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.iter, _, err = e.scan.idx.Seek(txn, []interface{}{seekVal})
+		if err != nil {
+			return nil, types.EOFAsNil(err)
+		}
+	}
+	for {
+		if e.finished {
+			return nil, nil
+		}
+		idxKey, h, err := e.iter.Next()
+		if err != nil {
+			return nil, types.EOFAsNil(err)
+		}
+		val := idxKey[0]
+		if !e.skipLowCmp {
+			cmp := indexCompare(val, e.lowVal)
```

Comment:
The seeked value may be less than low value, when column 'a' is of int type, and low value is a float value.